lib/linked_list.rb

```diff
@@ -7,39 +7,61 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity O(1)
+  # Space complexity O(1)
   def add_first(data)
-
+    a = Node.new(data)
+    a.next = @head
+    @head = a
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(0)????
   def get_first
-
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def length
-    return 0
+    length = 0
+    current = @head
+    until current.nil?
+      length += 1
+      current = current.next
+    end
+    return length
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
-
+    current = @head
+    until current.nil?
+      current = current.next
+    end
+    current = Node.new(data)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def get_last
-
+    current = @head
```

Comment:
Doesn't work, see below.

```suggestion
    return nil if @head.nil?
    current = @head
```